File: Java/ByRound/2153/2153_C_Symmetrical_Polygons.java
// Problem: CF 2153 C - Symmetrical Polygons
// https://codeforces.com/contest/2153/problem/C

/*
C. Symmetrical Polygons

Algorithms/Techniques: Sorting, Greedy, Polygon Construction

Time Complexity: O(n log n) due to sorting, where n is the number of sticks.
Space Complexity: O(n) for storing the array and auxiliary data structures.

Task:
Given n sticks with lengths a_i, find the maximum possible perimeter of a
non-degenerate, symmetrical, and strictly convex polygon that can be formed
using a subset of these sticks.

Key Observations:
- For a polygon to be convex and non-degenerate, the sum of any (n-1) sides must be greater than the remaining side.
- For symmetry, certain groupings of equal-length sides are required.
- An isosceles triangle (3 sides with two equal) can form a symmetrical convex polygon.
- We prioritize forming larger polygons (with more sides) for maximum perimeter.

Approach:
1. Sort the array to group identical elements together.
2. Process the sorted array to build unique side lengths and count pairs.
   - If two sticks have the same length, they can be used together as a symmetric pair.
3. Use greedy approach: Try to form polygons using as many sides as possible, ensuring that
   the sum of smaller sides is greater than the largest side (triangle inequality).
4. If at least 3 valid sides can be arranged into a symmetric polygon, return their total perimeter,
   otherwise return 0.
*/

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.util.*;
public class C {
    static BufferedReader br = new BufferedReader(
            new InputStreamReader(System.in));
    public static void main(String[] args) throws IOException {
        StringTokenizer s1
                = new StringTokenizer(br.readLine());
        int t = Integer.parseInt(s1.nextToken());
        while (t-- > 0) {
            StringTokenizer s2
                    = new StringTokenizer(br.readLine());
            int n = Integer.parseInt(s2.nextToken());
            int[] arr = new int[n];
            StringTokenizer s3
                    = new StringTokenizer(br.readLine());
            for (int i = 0; i < n; i++) arr[i] = Integer.parseInt(s3.nextToken());
            solve(arr, n);
        }
    }
    static void solve(int[] arr, int n) {
        Arrays.sort(arr); // O(n log n)
        long sum=0;
        int i=0,c=0;
        ArrayList<Integer> list=new ArrayList<>();
        while(i<n) {
            if(i==n-1 ||arr[i]!=arr[i+1]) list.add(arr[i++]); // Add single stick
            else if(arr[i]==arr[i+1]) { // If two sticks are same, pair them
                sum+=arr[i]+arr[i+1];
                i+=2;
                c+=2;
            }
        }
        int max1=0,max2=0;
        for(i=list.size()-1;i>0;i--) {
            if(sum+list.get(i-1)>list.get(i)) { // Check polygon inequality
                max1=list.get(i);
                max2=list.get(i-1);
                break;
            }
        }
        if(max1==0 &&!list.isEmpty() &&sum>list.getFirst()) max1=list.getFirst();  // If last element can form triangle with sum
        if(max1!=0) c++;
        if(max2!=0) c++;
        if(c<3) System.out.println(0);  // At least 3 sides needed to form polygon
        else System.out.println(sum+max1+max2);
    }
    static void print(int[] arr) {
        PrintWriter out=new PrintWriter(System.out);
        for (int i = 0; i < arr.length; i++)
            out.print(arr[i] + " ");
        out.print("\n");
        out.flush();
    }
}


// https://github.com/VaHiX/CodeForces/

File: CPP/ByRound/2153/2153_C_Symmetrical_Polygons.cpp
// Problem: CF 2153 C - Symmetrical Polygons
// https://codeforces.com/contest/2153/problem/C

/*
C. Symmetrical Polygons
Algorithms/Techniques: Sorting, Greedy, Two Pointers, Set Operations

Time Complexity: O(n log n) per test case due to sorting and linear passes.
Space Complexity: O(n) for storing the input and auxiliary arrays.

Problem Description:
We are given n sticks, and we want to choose a subset to form a symmetrical, strictly convex, and non-degenerate polygon.
The key insights are:
- For a polygon to be symmetrical, it needs at least one axis of symmetry.
- For a polygon to be convex and non-degenerate, the sum of any (n-1) sides must be greater than the largest side.
- We can build symmetrical polygons using pairs of equal-length sticks, and single sticks to balance.
- The approach focuses on grouping equal-length sticks, and then using greedy selection to achieve maximum perimeter under constraints.

Solution Approach:
1. Group sticks by their lengths.
2. Identify pairs (doubles) and singles (unique lengths).
3. For valid symmetrical constructions:
   - If there are no doubles, return 0.
   - Otherwise, consider adding doubles into the perimeter (double each to match symmetry).
   - Balance with singles for maximum perimeter while maintaining convexity.
4. Ensure the constructed polygon satisfies all required properties: non-degenerate, strictly convex, and symmetrical.
*/

#include <math.h>
#include <algorithm>
#include <iostream>
#include <iterator>
#include <string>
#include <utility>
#include <vector>

using namespace std;
#define R2D0 ios_base::sync_with_stdio(false), cin.tie(nullptr);
#define pb push_back
#define fi first
#define se second
#define mp make_pair
#define all(vec) vec.begin(), vec.end()
#define prinnt(x) cout << x << '\n'
#define cl(x, v) memset((x), (v), sizeof(x))
#define fo(i, n) for (int i = 0; i < n; ++i)
#define fore(i, k, n) for (int i = k; k < n ? i < n : i >= n; k < n ? ++i : --i)
#define fora(a) for (auto i : a)
#define tr(it, a) for (auto it = a.begin(); it != a.end(); ++it)
typedef long long ll;
typedef long double ld;
typedef pair<int, int> pii;
typedef pair<int, pii> piii;
typedef pair<ll, ll> pll;
typedef vector<int> vi;
typedef vector<ll> vl;
typedef vector<bool> vb;
typedef vector<pii> vpii;
typedef vector<pll> vpll;
typedef vector<vi> vvi;
typedef vector<vl> vvl;
const int N = 1e5 + 5, M = N;
const int MOD = 1e9 + 7, INF = 1e9 + 7;
const ll LINF = 1e18 + 7;
const ld EPS = 1e-9, PI = acos(-1.0);
template <typename T = int> istream &operator>>(istream &in, vector<T> &v) {
  for (auto &x : v)
    in >> x;
  return in;
}
template <typename T = int>
istream &operator>>(istream &in, vector<vector<T>> &v) {
  for (auto &x : v)
    in >> x;
  return in;
}
template <typename T = int>
ostream &operator<<(ostream &out, const vector<T> &v) {
  for (const T &x : v)
    out << x << ' ';
  return out;
}
template <typename A, typename B>
ostream &operator<<(ostream &os, const pair<A, B> &p) {
  return os << '(' << p.first << ", " << p.second << ')';
}
void swap(int &a, int &b) {
  a = a ^ b;
  b = a ^ b;
  a = a ^ b;
}
vector<string> RETU = {"NO\n", "YES\n", "-1\n"};
void solve() {
  int n, pivo, sinsize;
  ll ans = 0, othersum = 0;
  cin >> n;
  bool flagg = false;
  vi vec(n);
  vi dobles; // store pairs of equal lengths (doubles)
  vi singles; // store unique lengths
  cin >> vec;
  sort(all(vec));
  fo(i, n) {
    // Group equal lengths into pairs
    if (i < n - 1 && vec[i + 1] == vec[i]) {
      dobles.pb(vec[i]);
      i++; // Skip next element as paired with current
    } else {
      singles.pb(vec[i]);
    }
  }
  if (dobles.empty()) { // No pairs available
    cout << "0\n";
    return;
  }
  fora(dobles) { othersum += i; } // Sum of all paired sticks (doubles)
  othersum *= 2; // Each element in dobles contributes twice to total length (due to pairing)
  sort(all(singles));
  sinsize = singles.size();
  if (sinsize == 1 && singles[0] < othersum) { // Single stick smaller than sum of doubles
    ans += singles[0];
  } else if (sinsize > 1) {
    pivo = sinsize - 2; // Start from second-last element to look for optimal fit
    // Decrease pivo until (singles[pivo+1] - singles[pivo]) >= othersum,
    // or it means we can't use more pairs for symmetry
    while (pivo >= 0 && (singles[pivo + 1] - singles[pivo]) >= othersum)
      pivo--;
    if (pivo == (-1) && singles[0] < othersum) {
      ans += singles[0]; // Only first stick fits without violating symmetry
    } else if (pivo >= 0) {
      ans += singles[pivo + 1] + singles[pivo]; // Add two sticks for balance
    }
  }
  // If there are doubles or we have a valid contribution from singles,
  // add the total from doubles into result
  if (dobles.size() > 1 || ans != 0)
    ans += othersum;
  prinnt(ans);
}
int main() {
  R2D0;
  int tc = 1;
  cin >> tc;
  while (tc--)
    solve();
  return 0;
}


// https://github.com/VaHiX/CodeForces/